filament/src/fg2/details/DependencyGraph.h

```diff
@@ -0,0 +1,199 @@
+
+        //! Nodes can be moved
+        Node(Node&&) noexcept = default;
+
+        virtual ~Node() noexcept;
+
+        //! returns a unique id for this node
+        NodeID getId() const noexcept;
+
+        /** Prevents this node from being culled. Must be called before culling. */
+        void makeTarget() noexcept;
+
+        /** Returns true if this Node is a target */
+        bool isTarget() const noexcept;
+
+        /**
+         * Returns whether this node was culled.
+         * This is only valid after DependencyGraph::cull() has been called.
+         * @return true if the node has been culled, false otherwise.
+         */
+        bool isCulled() const noexcept;
+
+        /**
+         * return the reference count of this node. That is how many other nodes have links to us.
+         * This is only valid after DependencyGraph::cull() has been called.
+         * @return Number of nodes linking to this one.
+         */
+        uint32_t getRefCount() const noexcept;
+
+    public:
+        //! return the name of this node
+        virtual char const* getName() const noexcept;
+
+        //! called from DependencyGraph::cull() when a node a culled
+        virtual void onCulled(DependencyGraph* graph) noexcept;
+
+        //! output itself as a graphviz string
+        virtual utils::CString graphvizify() const noexcept;
+
+        //! output a graphviz color string for an Edge from this node
+        virtual utils::CString graphvizifyEdgeColor() const noexcept;
+
+    private:
+        // nodes that read from us: i.e. we have a reference to them
+        friend class DependencyGraph;
+        static const constexpr uint32_t TARGET = 0x80000000u;
+        uint32_t mRefCount = 0;     // how many references to us
+        const NodeID mId;           // unique id
+    };
+
+    using EdgeContainer = std::vector<Edge*>;
+    using NodeContainer = std::vector<Node*>;
+
+    /**
+     * Removes all edges and nodes from the graph.
+     * Note that these objects are not destroyed (DependencyGraph doesn't own them).
+     */
+    void clear() noexcept;
+
+    /** return the list of all edges */
+    EdgeContainer const& getEdges() const noexcept;
+
+    /** return the list of all nodes */
+    NodeContainer const& getNodes() const noexcept;
+
+    /**
+     * Returns the list of incoming edges to a node
+     * @param node the node to consider
+     * @return A list of incoming edges
+     */
+    EdgeContainer getIncomingEdges(Node const* node) const noexcept;
+
+    /**
+     * Returns the list of outgoing edges to a node
+     * @param node the node to consider
+     * @return A list of outgoing edges
+     */
+    EdgeContainer getOutgoingEdges(Node const* node) const noexcept;
+
+    Node const* getNode(NodeID id) const noexcept;
+
+    Node* getNode(NodeID id) noexcept;
+
+    //! cull unreferenced nodes. Links ARE NOT removed, only reference counts are updated.
+    void cull() noexcept;
```

Comment:
I'm having some trouble understanding the algorithm behind cull. When is a node culled? When there's no path from it to a "target" node?

Reply:
Yes excatly.